pkg/cmd/variable/set/set.go

```diff
@@ -161,6 +175,26 @@ func setRun(opts *SetOptions) error {
 		if err != nil {
 			return err
 		}
+
+		err = cmdutil.ValidateHasOnlyOneRemote(opts.HasRepoOverride, opts.Remotes)
+		if err != nil {
+			defaultRepo := cmdutil.DefaultRepo(opts.Remotes)
+
+			if defaultRepo != nil {
+				baseRepo = defaultRepo
+			} else if opts.Interactive {
+				selectedRepo, errSelectedRepo := cmdutil.PromptForRepo(baseRepo, opts.Remotes, opts.Prompter)
+
+				if errSelectedRepo != nil {
+					return errSelectedRepo
+				}
+
+				baseRepo = selectedRepo
+			} else {
+				return err
+			}
+		}
+
```

Comment:
**Discussing with @williammartin** ✨ 

1. Remove the line about `gh repo set-default` as this is technically the problem we're running into
2. Remove the line in `gh repo set-default` about this being used for repository + repository environment; actually should be called as NOT being impacted ...
3. In the remote prompt ordering, put the upstream first as we know upset users will want to hit enter
4. Call out potential follow up based on community interest to introduce a configuration option to avoid the prompting
5. Ensure release notes

Reply:
> 1. If a user specifies `--owner` this is an organization secret; otherwise it is a repository secret of some sort
I think you meant `--org` instead of `--owner`, if so ✅ 

> 2. If a user specifies `--repo` this has the highest precedence because the user is explicitly telling us what to use
I think number 2 would be a secret or variable set with the `--env` flag, to set variables and secrets on a deployment environment level. which I think it be place between `--org` and `--repo` in terms of scope level

Reply:
> During our conversation, Will highlighted several things including:
> 
> 1. Remove the line about `gh repo set-default` as this is technically the problem we're running into

OK, reverted the last commit, it will still select the default in the select prompt if there is a default but submitting is up to the user

> 2. This PR should explicitly call out in `gh repo set-default` that it IS NOT used for repository + repository environment

Idd, it would do the exact opposite to what's in the docs, so the guidelines for `gh repo set-default` should change as well, added a proposal. Let me know if you see any other places where we can add this note

> 3. In the remote prompt ordering, the upstream should be first as we know upset users will want to hit enter

We piggyback on `Remotes` which is doing sorting by itself using the `Less` function and the `sort` library from golang 👉 https://github.com/cli/cli/blob/95a2f95f75f4b143699d87294788210ffb558248/context/remote.go#L60-L78 so the order is always `upstream` first even though the default selected option might be the `origin` one

> 4. With these changes, we need to ensure it is called out in release notes and mention the community should follow up in this issue about interest in a configuration option

👍 

> 5. Will also called out a larger problem regarding the inconsistent handling of repository defaults that might necessitate further changes

I agree, I also mentioned something similar here 👉  https://github.com/cli/cli/pull/9083#discussion_r1701547993. problem is, it's not just a feature or a change to a command, it's refactoring with a higher level overview in mind. Is that something I can kickstart in https://github.com/cli/cli/discussions/categories/ideas? Or do you prefer to have this discussion internally first?